lib/services/local/saml_idp_service_provider.go

```diff
@@ -295,3 +307,99 @@ func (s *SAMLIdPServiceProviderService) generateAndSetEntityDescriptor(sp types.
+	// value greater than 0. Otherwise, simply append to ed.SPSSODescriptors.
+	if embeddedSPSSODescriptorIndex == 0 {
+		ed.SPSSODescriptors = append(ed.SPSSODescriptors, newEmbeddedSPSSODescriptor)
+	} else {
+		ed.SPSSODescriptors[embeddedSPSSODescriptorIndex] = newEmbeddedSPSSODescriptor
+	}
+
+	edWithAttributes, err := xml.MarshalIndent(ed, " ", "    ")
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	s.log.Infof("attribute mapping embedded in entity descriptor")
+	sp.SetEntityDescriptor(string(edWithAttributes))
+	return nil
+}
+
+// spSSODescriptorWithRequestedAttributes builds new saml.SPSSODescriptor populated with
+// types.SAMLAttributeMapping (attributeMapping input) converted to saml.RequestedAttributes format.
+func spSSODescriptorWithRequestedAttributes(attributeMapping []*types.SAMLAttributeMapping) saml.SPSSODescriptor {
+	var reqs []saml.RequestedAttribute
+	for _, v := range attributeMapping {
+		var reqAttrs saml.RequestedAttribute
+		reqAttrs.FriendlyName = v.Name
+		reqAttrs.Name = v.Name
+		reqAttrs.NameFormat = v.NameFormat
+		reqAttrs.Values = append(reqAttrs.Values, saml.AttributeValue{
+			Value: v.Value,
+		})
+		reqs = append(reqs, reqAttrs)
+	}
+
+	return saml.SPSSODescriptor{
+		AttributeConsumingServices: []saml.AttributeConsumingService{
+			{
+				// ServiceNames is hardcoded with value TELEPORT_SAML_IDP to make the descriptor
+				// recognizable throughout SAML SSO flow. Attribute mapping should only ever
+				// edit SPSSODescriptor containing TELEPORT_SAML_IDP element. Otherwise, we risk
+				// overriding SP managed SPSSODescriptor!
+				ServiceNames:        []saml.LocalizedName{{Value: teleportSAMLIdP}},
+				RequestedAttributes: reqs,
+			},
+		},
+	}
+}
+
+// samlSPSSODescriptorToProtoSAMLAttributeMapping converts []saml.SPSSODescriptor to []*types.SAMLAttributeMapping
+// and returns SPSSODescriptor element index and converted attribute mapping values.
+// The correct SPSSODescriptor is determined by searching for AttributeConsumingService element with ServiceNames named
+// TELEPORT_SAML_IDP.
+func samlSPSSODescriptorToProtoSAMLAttributeMapping(spSSODescriptors []saml.SPSSODescriptor) (int, []*types.SAMLAttributeMapping) {
+	embeddedSPSSODescriptorIndex := 0
+	attrs := make([]*types.SAMLAttributeMapping, 0)
+	for descriptorIndex, descriptor := range spSSODescriptors {
+		for _, acs := range descriptor.AttributeConsumingServices {
+			for _, serviceName := range acs.ServiceNames {
+				if serviceName.Value == teleportSAMLIdP {
+					embeddedSPSSODescriptorIndex = descriptorIndex
+					for _, reqAttr := range acs.RequestedAttributes {
+						var embeddedAttribute types.SAMLAttributeMapping
+						for _, reqAttrVal := range reqAttr.Values {
+							embeddedAttribute.Name = reqAttr.Name
+							embeddedAttribute.NameFormat = reqAttr.NameFormat
+							embeddedAttribute.Value = reqAttrVal.Value
+						}
+						attrs = append(attrs, &embeddedAttribute)
```

Comment:
This can be written slightly more efficiently, without an extra variable:

```go
for _, reqAttrVal := range reqAttr.Values {
	attrs = append(attrs, &types.SAMLAttributeMapping{
		Name: reqAttr.Name,
		NameFormat: reqAttr.NameFormat,
		Value: retAttr.Value,
	}
}
```